ext/RMagick/rmimage.c

```diff
@@ -10949,6 +10949,8 @@ Image_recolor(VALUE self, VALUE color_matrix)
 #endif
 
     image = rm_check_destroyed(self);
+    color_matrix = rb_Array(color_matrix);
```

Comment:
Should we raise an error here if it's not an array?

```c
if (!RB_TYPE_P(color_matrix, T_ARRAY))
{
    rb_raise(rb_eTypeError, "color_matrix must be of type Array");
}
```

I guess that won't work for types that can be converted to array, but it seems easy for the user to handle that, and helps them find bugs faster.

Reply:
https://github.com/rmagick/rmagick/blob/c0fc7b4e983cb6bb6e5c7df0184d0e01aa9e7dce/ext/RMagick/rmilist.c#L864-L867

In this part, the internal data clearly was broken. The exception should be raised immediately.

----

Here is explanation of this pull request.

I love Ruby's ideology of duck typing. 
So, I think that method should accept the any argument as possible.
If we use `rb_Array` in here, the argument object will be convert implicitly if the object has `to_a` method.

```
require 'rmagick'

data = 'foobarbaz'
def data.to_a
  [1, 2, 3, 4]
end

image = Magick::Image.read('sample.png')[0]
p image.recolor(data).write('test.png')
```

Reply:
@Watson1978 I'm open to the duck typing approach, but I don't think this is using `to_a`, I think it's using `Array`, which calls `to_a` if it exists, but otherwise just puts the object in an array:

```irb
> 'x'.to_a
Traceback (most recent call last):
        2: from /home/fletch/.rvm/rubies/ruby-2.5.3/bin/irb:11:in `<main>'
        1: from (irb):1
NoMethodError (undefined method `to_a' for "x":String)
Did you mean?  to_r
               to_f
               to_i
               to_s
               to_c
> Array('x')
 => ["x"]
```

`Array` will convert basically anything to an array whether it has a `to_a` method or not:

```irb
> Array(nil)
 => []
> Array(Object.new)
 => [#<Object:0x00000000011511b0>]
```

Can we change it to call `to_a`? Then the user will get an earlier and hopefully clearer error when they pass in invalid input.

Reply:
We already using `rb_Array` in some places. Did you mean we have to change all?

Reply:
Hmm, good question. I don't really know about the other places. We may want to look at them each separately. For the case of `color_matrix`, I can't think of a scenario where it makes sense to have something that does not have a `to_a` method, since it should always come through as a nested array (`[[1,2,3], [4,5,6]]`), or something that converts to it.

It looks like `Kernel#Array` is the same as `rb_Array()` based on [the source code](https://github.com/ruby/ruby/blob/7fa16fd9624ca38f2d82c121a9af5f22c87e2697/object.c#L3764-L3779). It looks like it wraps the object in a new array if it doesn't respond to `to_a`:

```c
tmp = rb_check_to_array(val);
if (NIL_P(tmp)) {
    return rb_ary_new3(1, val);
}
```

I also ran your test code above to confirm and it raised a `TypeError` later in the process, when it should be a `NoMethodError`:

```irb
> image.recolor('x')
TypeError: no implicit conversion to float from string
```

Reply:
Side note, it seems really strange to me that `nil` has a `to_a` method:

```irb
> nil.to_a
=> []
```

Reply:
Actually, I'm wondering if we should use `.to_ary` instead. The semantics are slightly different, but it seems a little more strict. I'm looking at the ruby source, and I'm not sure how we would call either `.to_a` or `.to_ary` in C.

Reply:
Make sense. We should use `rb_check_array_type()` Ruby C API instead of `rb_Array()`.

Reply:
ah, `rb_check_array_type()` does not raise exception if object does not has `to_ary`

Reply:
For more strict, we need to add 

```
if (NIL_P(v)) rb_raise(rb_eArgError, "invalid argument");
```

before `rb_Array()`.

Reply:
@Watson1978 It looks like maybe we could do:

```c
color_matrix = rb_convert_type(color_matrix, T_ARRAY, "Array", "to_ary");
```

I don't know what's the best option here. Unfortunately, it doesn't look like Ruby provides a good C API for strict conversion to an array. We can leave it as `rb_Array` for now if you think that is better. Maybe we can revisit it at another point. What do you think?

Reply:
`rb_check_array_type()` API just return `nil` if fail the conversion.
Now, I made the utility function using `rb_check_array_type()`.

Reply:
Excellent! Thanks so much for your patience on this.